![Cabec%CC%A7alho_notebook.png](cabecalho_notebook.png)

# PCA - Tarefa 01: *HAR* com PCA

Vamos trabalhar com a base da demonstração feita em aula, mas vamos explorar um pouco melhor como é o desempenho da árvore variando o número de componentes principais.

In [3]:
import pandas as pd
import matplotlib.pyplot as plt

from sklearn.tree import DecisionTreeClassifier

from sklearn.decomposition import PCA
from sklearn.metrics import accuracy_score
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import GridSearchCV

filename_features = "UCI HAR Dataset/features.txt"
filename_labels = "UCI HAR Dataset/activity_labels.txt"

filename_subtrain = "UCI HAR Dataset/train/subject_train.txt"
filename_xtrain = "UCI HAR Dataset/train/X_train.txt"
filename_ytrain = "UCI HAR Dataset/train/y_train.txt"

filename_subtest = "UCI HAR Dataset/test/subject_test.txt"
ffilename_xtest = "UCI HAR Dataset/test/X_test.txt"
filename_ytest = "UCI HAR Dataset/test/y_test.txt"

features = pd.read_csv(filename_features, header=None, names=['nome_var'], sep="#").squeeze()
labels = pd.read_csv(filename_labels, sep='\\s+', header=None, names=['cod_label', 'label'])

subject_train = pd.read_csv(filename_subtrain, header=None, names=['subject_id']).squeeze()
X_train = pd.read_csv(filename_xtrain, sep='\\s+', header=None, names=features.tolist())
y_train = pd.read_csv(filename_ytrain, header=None, names=['cod_label'])

subject_test = pd.read_csv(filename_subtest, header=None, names=['subject_id']).squeeze()
X_test = pd.read_csv(ffilename_xtest,sep='\\s+', header=None, names=features.tolist())
y_test = pd.read_csv(filename_ytest, header=None, names=['cod_label'])

## Árvore de decisão

Rode uma árvore de decisão com todas as variáveis, utilizando o ```ccp_alpha=0.001```. Avalie a acurácia nas bases de treinamento e teste. Avalie o tempo de processamento.

In [5]:
%%time
#Arvore de Decisao
clf = DecisionTreeClassifier(random_state=1234, ccp_alpha=0.001).fit(X_train, y_train)

# previsões
y_train_pred = clf.predict(X_train)
y_test_pred = clf.predict(X_test)

# Calculo da Acuracia
acc_train = accuracy_score(y_train, y_train_pred)
acc_test = accuracy_score(y_test, y_test_pred)

print(f'Acurácia da melhor árvore na base de treino:    {acc_train*100:.1f}')
print(f'Acurácia da melhor árvore na base de teste:     {acc_test*100:.1f}')

Acurácia da melhor árvore na base de treino:    97.6
Acurácia da melhor árvore na base de teste:     88.0
CPU times: total: 5.23 s
Wall time: 5.24 s


## Árvore com PCA

Faça uma análise de componemtes principais das variáveis originais. Utilize apenas uma componente. Faça uma árvore de decisão com esta componente como variável explicativa.

- Avalie a acurácia nas bases de treinamento e teste
- Avalie o tempo de processamento

In [7]:
%%time
prcomp = PCA(n_components=1).fit(X_train)

pc_train = prcomp.transform(X_train)
pc_test = prcomp.transform(X_test)

# Arvore com pca
clf_pca = DecisionTreeClassifier(random_state=1234, ccp_alpha=0.001).fit(pc_train, y_train)

# Previsoes pca
y_train_pred_pca = clf_pca.predict(pc_train)
y_test_pred_pca = clf_pca.predict(pc_test)

#Acuracia pca
acc_train_pca = accuracy_score(y_train, y_train_pred_pca)
acc_test_pca = accuracy_score(y_test, y_test_pred_pca)

print(f'Acurácia da melhor árvore com PCA na base de treino:    {acc_train_pca*100:.1f}')
print(f'Acurácia da melhor árvore com PCA na base de teste:     {acc_test_pca*100:.1f}')


Acurácia da melhor árvore com PCA na base de treino:    50.0
Acurácia da melhor árvore com PCA na base de teste:     45.7
CPU times: total: 594 ms
Wall time: 170 ms


## Testando o número de componentes

Com base no código acima, teste a árvore de classificação com pelo menos as seguintes possibilidades de quantidades de componentes: ```[1, 2, 5, 10, 50]```. Avalie para cada uma delas:

- Acurácia nas bases de treino e teste
- Tempo de processamento


In [9]:
%%time
prcomp = PCA(n_components=2).fit(X_train)

pc_train = prcomp.transform(X_train)
pc_test = prcomp.transform(X_test)

# Arvore com pca
clf_pca = DecisionTreeClassifier(random_state=1234, ccp_alpha=0.001).fit(pc_train, y_train)

# Previsoes pca
y_train_pred_pca = clf_pca.predict(pc_train)
y_test_pred_pca = clf_pca.predict(pc_test)

#Acuracia pca
acc_train_pca = accuracy_score(y_train, y_train_pred_pca)
acc_test_pca = accuracy_score(y_test, y_test_pred_pca)

print(f'Acurácia da melhor árvore com PCA na base de treino:    {acc_train_pca*100:.1f}')
print(f'Acurácia da melhor árvore com PCA na base de teste:     {acc_test_pca*100:.1f}')

Acurácia da melhor árvore com PCA na base de treino:    61.3
Acurácia da melhor árvore com PCA na base de teste:     58.5
CPU times: total: 500 ms
Wall time: 122 ms


In [10]:
%%time
prcomp = PCA(n_components=5).fit(X_train)

pc_train = prcomp.transform(X_train)
pc_test = prcomp.transform(X_test)

# Arvore com pca
clf_pca = DecisionTreeClassifier(random_state=1234, ccp_alpha=0.001).fit(pc_train, y_train)

# Previsoes pca
y_train_pred_pca = clf_pca.predict(pc_train)
y_test_pred_pca = clf_pca.predict(pc_test)

#Acuracia pca
acc_train_pca = accuracy_score(y_train, y_train_pred_pca)
acc_test_pca = accuracy_score(y_test, y_test_pred_pca)

print(f'Acurácia da melhor árvore com PCA na base de treino:    {acc_train_pca*100:.1f}')
print(f'Acurácia da melhor árvore com PCA na base de teste:     {acc_test_pca*100:.1f}')

Acurácia da melhor árvore com PCA na base de treino:    84.6
Acurácia da melhor árvore com PCA na base de teste:     78.9
CPU times: total: 484 ms
Wall time: 124 ms


In [11]:
%%time
prcomp = PCA(n_components=10).fit(X_train)

pc_train = prcomp.transform(X_train)
pc_test = prcomp.transform(X_test)

# Arvore com pca
clf_pca = DecisionTreeClassifier(random_state=1234, ccp_alpha=0.001).fit(pc_train, y_train)

# Previsoes pca
y_train_pred_pca = clf_pca.predict(pc_train)
y_test_pred_pca = clf_pca.predict(pc_test)

#Acuracia pca
acc_train_pca = accuracy_score(y_train, y_train_pred_pca)
acc_test_pca = accuracy_score(y_test, y_test_pred_pca)

print(f'Acurácia da melhor árvore com PCA na base de treino:    {acc_train_pca*100:.1f}')
print(f'Acurácia da melhor árvore com PCA na base de teste:     {acc_test_pca*100:.1f}')

Acurácia da melhor árvore com PCA na base de treino:    89.3
Acurácia da melhor árvore com PCA na base de teste:     82.4
CPU times: total: 750 ms
Wall time: 192 ms


In [12]:
%%time
prcomp = PCA(n_components=50).fit(X_train)

pc_train = prcomp.transform(X_train)
pc_test = prcomp.transform(X_test)

# Arvore com pca
clf_pca = DecisionTreeClassifier(random_state=1234, ccp_alpha=0.001).fit(pc_train, y_train)

# Previsoes pca
y_train_pred_pca = clf_pca.predict(pc_train)
y_test_pred_pca = clf_pca.predict(pc_test)

#Acuracia pca
acc_train_pca = accuracy_score(y_train, y_train_pred_pca)
acc_test_pca = accuracy_score(y_test, y_test_pred_pca)

print(f'Acurácia da melhor árvore com PCA na base de treino:    {acc_train_pca*100:.1f}')
print(f'Acurácia da melhor árvore com PCA na base de teste:     {acc_test_pca*100:.1f}')

Acurácia da melhor árvore com PCA na base de treino:    91.9
Acurácia da melhor árvore com PCA na base de teste:     82.3
CPU times: total: 1.47 s
Wall time: 687 ms


## Conclua

- O que aconteceu com a acurácia?
- O que aconteceu com o tempo de processamento?

**Comparativo de Acuracia**
n_componentes|Acuracia Treino %|Acuracia Teste %
-------------|-----------------|----------------
1|50.0|45.7
2|61.3|58.5
5|84.6|78.9
10|89.3|82.4
50|91.9|82.3


- Com 1 ou 2 componentes a acuracia é baixa, indicando que muita informacao foi perdida na redução da dimensionalidade (PCA)
- Com 5 componentes a acurácia aumenta bastante e com 10 componentes a acurácia continua subindo
- Com 50 componentes, a acurácia melhora no treino mas abaixa um pouco no teste, indicando que adicionar mais componentes por causar overfitting

**Comparativo de Tempo de Processamento**
n_componentes|CPU Time|Wall Time 
-------------|--------|----------
1|594 ms|170 ms
2|500 ms|122 ms
5|484 ms|124 ms
10|750 ms|192 ms
50|1.47 s|687 ms

- Com 1, 2 e 5 componentes, o tempo de processamento é baixo e estável, mostrando que o modelo consegue processar rapidamente com poucas variaveis
- Com 10 componentes há um aumento no tempo, porém ainda numa velocidade razoavel
- Com 50 componentes o tempo quase dobra em comparação a 10 componentes

**Conclusão**
- Sugere-se manter entre 5 a 10 componentes para um equilibrio entre acurácia e eficiencia em tempo de processamento.